src/pkcs15init/pkcs15-lib.c

```diff
@@ -1547,37 +1604,50 @@ sc_pkcs15init_generate_key(struct sc_pkcs15_card *p15card, struct sc_profile *pr
 	pubkey = &pubkey_args.key;
 	if (!pubkey->alg_id)   {
 		pubkey->alg_id = calloc(1, sizeof(struct sc_algorithm_id));
-		if (!pubkey->alg_id)
+		if (!pubkey->alg_id) {
+			if (r < 0) {
+				free(pubkey_args.key.u.ec.params.der.value);
+				free(pubkey_args.key.u.ec.params.named_curve);
+				sc_pkcs15_free_object(object);
+			}
 			LOG_FUNC_RETURN(ctx, SC_ERROR_OUT_OF_MEMORY);
+		}
 
 		sc_init_oid(&pubkey->alg_id->oid);
 		pubkey->alg_id->algorithm = pubkey->algorithm;
 	}
 
 	pubkey_args.id = key_info->id;
 	r = sc_pkcs15_encode_pubkey(ctx, pubkey, &object->content.value, &object->content.len);
-	LOG_TEST_RET(ctx, r, "Failed to encode public key");
+	LOG_TEST_GOTO_ERR(ctx, r, "Failed to encode public key");
 
 	r = sc_pkcs15init_add_object(p15card, profile, SC_PKCS15_PRKDF, object);
-	LOG_TEST_RET(ctx, r, "Failed to add generated private key object");
+	LOG_TEST_GOTO_ERR(ctx, r, "Failed to add generated private key object");
 
 	if (!r && profile->ops->emu_store_data)   {
 		r = profile->ops->emu_store_data(p15card, profile, object, NULL, NULL);
 		if (r == SC_ERROR_NOT_IMPLEMENTED)
 			r = SC_SUCCESS;
-		LOG_TEST_RET(ctx, r, "Card specific 'store data' failed");
+		LOG_TEST_GOTO_ERR(ctx, r, "Card specific 'store data' failed");
 	}
 
 	r = sc_pkcs15init_store_public_key(p15card, profile, &pubkey_args, NULL);
-	LOG_TEST_RET(ctx, r, "Failed to store public key");
+	if (r < 0)
+		sc_pkcs15_remove_object(p15card, object);
+	LOG_TEST_GOTO_ERR(ctx, r, "Failed to store public key");
 
 	if (res_obj)
 		*res_obj = object;
+	object = NULL;
```

Comment:
Namely in `sc_pkcs15init_generate_key()`, in `sc_pkcs15init_add_object()` the ownership of object was moved to `p15card` by [`sc_pkcs15_add_object`](https://github.com/OpenSC/OpenSC/blob/afab2e6fadc20c925c2f32a8c059316583d6a562/src/pkcs15init/pkcs15-lib.c#L3204) and therefore should not be freed, I think. In case of failure, there are `LOG_TEST_GOTO_ERR` that go straight to `err`.
But I will check once more as well as the other places.